### ISSUE TITLE
Feature request: Maximum and minimum.

### ISSUE THREAD

**Issue by DavidMikeSimon** [magnusvk/counter_culture#117]
When `:delta_column` is provided, then `SUM` is used instead of `COUNT`. What if other composable SQL aggregate functions were available as well? In particular, `MIN` and `MAX` would be very useful for me.

For example, suppose we want to track in each Category the price of its most expensive item.

```ruby
class Product < ActiveRecord::Base
  belongs_to :category
  counter_cache :products, :column_name => 'product_max_price', :max_of_column => 'price'
end

class Category
  has_many :products
end
```

If a Product's price is changed to P and it belongs to Category C, counter_cache runs this query:

```SQL
UPDATE categories SET product_price_max = P WHERE id = C AND product_price_max < P
```

The only hitch to this is: If a product changes categories, the database would be have to rescan all products in the old category to get the correct value.

Would a PR to add this feature be accepted?

**Comment by DavidMikeSimon** [comment]
Argh, thinking about this further, the UPDATE query above wouldn't work if P is already the most expensive product and its price decreases. Since the optimization would only work when the value increases, it's probably not worth the trouble.

However, there is another way to speed up the calculation: create an index in products on `(category_id, price)`.

**Comment by magnusvk** [comment]
This is very similar to #115, right? The problem I see is the following: Both `delta_column` support and the simple counter cache allow us to issue update statements without having to load and evaluate the current cached value or, worse, all values in the table. If you wanted to keep a min and max up to date, you would have to always compare the value of the record in hand to the current cached value to determine if you want to update it. An average would be slightly more complicated still -- you'd need at least the current average and count to re-calculate an average on adding / removing values.

I frankly think that's a substantial amount of complexity that may be more easily handled outside of counter_culture? Feel free to submit a PR that proves me wrong though and I'm happy to take a look. :)

**Comment by DavidMikeSimon** [comment]
I agree on the database overhead; there's not much way to get around that.

But I don't think much code complexity would be added; counter_cache is already watching for all the scenarios that would necessitate updating the cached value.

**Comment by magnusvk** [comment]
Yeah, you may be right. If so, I'd love to look at a PR and get that merged. :)

**Comment by DavidMikeSimon** [comment]
@magnusvk I'm curious; why does counter_culture run the updates after the transaction commits?

**Comment by magnusvk** [comment]
Because that avoids deadlock problems in Postgres, [see here](http://mina.naguib.ca/blog/2010/11/22/postgresql-foreign-key-deadlocks.html). Arguably that behavior should be optional.

**Comment by DavidMikeSimon** [comment]
According to that article and the Postgres release notes, it actually looks like this kind of deadlock was fixed in Postgres 9.3, which came out a couple years ago.

**Comment by magnusvk** [comment]
You are right. I submitted #120 to make that optional behavior.
